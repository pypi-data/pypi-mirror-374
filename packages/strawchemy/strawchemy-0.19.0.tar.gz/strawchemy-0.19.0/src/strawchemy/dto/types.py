"""DTO domain types."""

from __future__ import annotations

import dataclasses
from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any, Literal, Optional, Union, final, get_type_hints

from typing_extensions import TypeAlias, override

from strawchemy.utils import get_annotations

if TYPE_CHECKING:
    from collections.abc import Callable, Mapping


__all__ = ("DTOAuto", "DTOConfig", "DTOFieldConfig", "DTOMissing", "ExcludeFields", "IncludeFields", "Purpose")

DTOScope: TypeAlias = Literal["global", "dto"]


IncludeFields: TypeAlias = Union[list[str], set[str], Literal["all"]]
ExcludeFields: TypeAlias = Union[list[str], set[str]]


@final
class DTOMissing:
    """A sentinel type to detect if a parameter is supplied or not when.

    constructing pydantic FieldInfo.
    """


@final
class DTOAuto: ...


@final
class DTOSkip: ...


@final
class DTOUnset:
    @override
    def __str__(self) -> str:
        return ""

    @override
    def __repr__(self) -> str:
        return "DTOUnset"

    def __bool__(self) -> bool:
        return False


class Purpose(str, Enum):
    """For identifying the purpose of a DTO to the factory.

    The factory will exclude fields marked as private or read-only on the domain model depending
    on the purpose of the DTO.

    Example:
    ```python
    ReadDTO = dto.factory("AuthorReadDTO", Author, purpose=dto.Purpose.READ)
    ```
    """

    READ = "read"
    """To mark a DTO that is to be used to serialize data returned to
    clients."""
    WRITE = "write"
    """To mark a DTO that is to deserialize and validate data provided by
    clients."""
    COMPLETE = "complete"
    """To mark a DTO that is to deserialize and validate data provided by
    clients. Fields marked as TO_COMPLETE must not be null."""


@dataclass
class PurposeConfig:
    """Mark the field as read-only, or private."""

    type_override: Optional[Any] = DTOMissing
    validator: Optional[Callable[[Any], Any]] = None
    """Single argument callables that are defined on the DTO as validators for the field."""
    alias: Optional[str] = None
    """Customize name of generated DTO field."""
    partial: Optional[bool] = None


@dataclass
class DTOFieldConfig:
    """For configuring DTO behavior on SQLAlchemy model fields."""

    purposes: set[Purpose] = field(default_factory=lambda: {Purpose.READ, Purpose.WRITE})
    default_config: PurposeConfig = field(default_factory=PurposeConfig)
    configs: dict[Purpose, PurposeConfig] = field(default_factory=dict)

    def purpose_config(self, dto_config: DTOConfig) -> PurposeConfig:
        return self.configs.get(dto_config.purpose, self.default_config)


@dataclass
class DTOConfig:
    """Control the generated DTO.

    This class holds configuration settings that influence how a Data Transfer
    Object (DTO) is generated by the DTO factory. It allows customization of
    field inclusion/exclusion, optionality, type hints, and field aliasing based
    on the intended purpose (read, write, etc.) of the DTO.

    Attributes:
        purpose: Configure the DTO for "read", "write", or "complete" operations.
            Determines which fields from the source model are included based on
            their `DTOFieldConfig`.
        include: Explicitly include fields from the source model in the generated
            DTO. Can be a list or set of field names, or the literal "all" to
            include all fields not explicitly excluded. Defaults to an empty set.
        exclude: Explicitly exclude fields from the source model. Can be a list
            or set of field names. Defaults to an empty set. Setting this
            implicitly sets `include` to "all".
        partial: If True, makes all fields in the generated DTO optional.
            Defaults to None.
        partial_default: The default value assigned to fields when `partial` is
            True and the field is not provided. Defaults to None.
        unset_sentinel: A sentinel object used to represent fields that are not
            set, particularly useful when distinguishing between a field explicitly
            set to `None` and a field that was not provided at all. Defaults to
            `DTO_UNSET`.
        type_overrides: A mapping to override the type annotations for specific
            fields in the generated DTO. Keys can be field names or types,
            values are the overriding types. Defaults to an empty dict.
        annotation_overrides: A dictionary to directly set or override the type
            annotations for specific fields by name. Defaults to an empty dict.
        aliases: A mapping of source model field names to their desired names
            (aliases) in the generated DTO. Defaults to an empty dict. Mutually
            exclusive with `alias_generator`.
        alias_generator: A callable that accepts a field name and returns its
            alias for the generated DTO. Defaults to None. Mutually exclusive
            with `aliases`.

    Raises:
        ValueError: If both `aliases` and `alias_generator` are provided, or
            if `exclude` is set while `include` is also set to a specific list/set
            (i.e., not "all" or empty).
    """

    purpose: Purpose
    """Configure the DTO for "read" or "write" operations."""
    include: IncludeFields = field(default_factory=set)
    """Explicitly include fields from the generated DTO."""
    exclude: ExcludeFields = field(default_factory=set)
    """Explicitly exclude fields from the generated DTO. Implies `include="all"`."""
    partial: Optional[bool] = None
    """Make all field optional."""
    partial_default: Any = None
    unset_sentinel: Any = DTOUnset
    type_overrides: Mapping[Any, Any] = field(default_factory=dict)
    annotation_overrides: dict[str, Any] = field(default_factory=dict)
    aliases: Mapping[str, str] = field(default_factory=dict)
    exclude_defaults: bool = False
    alias_generator: Optional[Callable[[str], str]] = None
    scope: Optional[DTOScope] = None
    exclude_from_scope: bool = False
    tags: set[str] = field(default_factory=set)

    def __post_init__(self) -> None:
        if self.aliases and self.alias_generator is not None:
            msg = "You must set `aliases` or `alias_generator`, not both"
            raise ValueError(msg)
        if self.include and self.include != "all" and self.exclude:
            msg = "When using `exclude` you must set `include='all' or leave it unset`"
            raise ValueError(msg)
        if self.exclude:
            self.include = "all"

    def copy_with(
        self,
        purpose: Union[Purpose, type[DTOUnset]] = DTOUnset,
        include: Optional[IncludeFields] = None,
        exclude: Optional[ExcludeFields] = None,
        partial: Union[bool, None, type[DTOUnset]] = DTOUnset,
        unset_sentinel: Union[Any, type[DTOUnset]] = DTOUnset,
        type_overrides: Union[Mapping[Any, Any], type[DTOUnset]] = DTOUnset,
        annotation_overrides: Union[dict[str, Any], type[DTOUnset]] = DTOUnset,
        aliases: Union[Mapping[str, str], type[DTOUnset]] = DTOUnset,
        exclude_defaults: Union[bool, type[DTOUnset]] = DTOUnset,
        alias_generator: Union[Callable[[str], str], type[DTOUnset]] = DTOUnset,
        partial_default: Union[Any, type[DTOUnset]] = DTOUnset,
        scope: Union[DTOScope, type[DTOUnset]] = DTOUnset,
        exclude_from_scope: Union[bool, type[DTOUnset]] = DTOUnset,
        tags: Union[set[str], type[DTOUnset]] = DTOUnset,
    ) -> DTOConfig:
        """Create a copy of the DTOConfig with the specified changes."""
        if include is None and exclude is None:
            include, exclude = self.include, self.exclude
        else:
            include = include or []
            exclude = exclude or []
        return DTOConfig(
            include=include,
            exclude=exclude,
            purpose=self.purpose if purpose is DTOUnset else purpose,
            partial=self.partial if partial is DTOUnset else partial,
            unset_sentinel=self.unset_sentinel if unset_sentinel is DTOUnset else unset_sentinel,
            type_overrides=self.type_overrides if type_overrides is DTOUnset else type_overrides,
            annotation_overrides=self.annotation_overrides
            if annotation_overrides is DTOUnset
            else annotation_overrides,
            aliases=self.aliases if aliases is DTOUnset else aliases,
            exclude_defaults=self.exclude_defaults if exclude_defaults is DTOUnset else exclude_defaults,
            alias_generator=self.alias_generator if alias_generator is DTOUnset else alias_generator,
            partial_default=self.partial_default if partial_default is DTOUnset else partial_default,
            scope=self.scope if scope is DTOUnset else scope,
            exclude_from_scope=self.exclude_from_scope if exclude_from_scope is DTOUnset else exclude_from_scope,
            tags=self.tags if tags is DTOUnset else tags,
        )

    def with_base_annotations(self, base: type[Any]) -> DTOConfig:
        """Merge type annotations from a base class into this DTOConfig.

        Args:
            base: The base class to extract type annotations from

        Returns:
            A new DTOConfig instance with:
            - Type annotations from the base class merged into annotation_overrides
            - Updated include set to include all fields if exclude is specified or include was "all"

        The method handles two cases:
        1. When include is "all" or exclude is specified: All fields from the base class are included
        2. When specific fields are included: Only those fields are added to the include set
        """
        include: set[str] = set(self.include) if self.include != "all" else set()
        include_all = self.include == "all" or self.exclude
        annotation_overrides: dict[str, Any] = self.annotation_overrides
        try:
            base_annotations = get_type_hints(base, include_extras=True)
        except NameError:
            base_annotations = get_annotations(base)
        for name, annotation in base_annotations.items():
            if not include_all:
                include.add(name)
            annotation_overrides[name] = annotation
        return dataclasses.replace(
            self,
            include="all" if include_all else include,
            annotation_overrides=annotation_overrides,
        )

    def alias(self, name: str) -> Optional[str]:
        if self.aliases:
            return self.aliases.get(name)
        if self.alias_generator is not None:
            return self.alias_generator(name)
        return None
