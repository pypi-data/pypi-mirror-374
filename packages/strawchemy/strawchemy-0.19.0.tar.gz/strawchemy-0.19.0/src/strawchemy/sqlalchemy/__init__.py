"""SQLAlchemy integration for Strawchemy.

This package provides SQLAlchemy-based repository implementations for use with
Strawberry GraphQL.
"""

from __future__ import annotations

from .repository import SQLAlchemyGraphQLAsyncRepository, SQLAlchemyGraphQLRepository, SQLAlchemyGraphQLSyncRepository

__all__ = ("SQLAlchemyGraphQLAsyncRepository", "SQLAlchemyGraphQLRepository", "SQLAlchemyGraphQLSyncRepository")
