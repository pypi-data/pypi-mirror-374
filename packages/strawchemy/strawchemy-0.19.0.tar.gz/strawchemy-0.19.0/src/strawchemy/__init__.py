"""Custom DTO implementation."""

from __future__ import annotations

from .config.base import StrawchemyConfig
from .mapper import Strawchemy
from .sqlalchemy.hook import QueryHook
from .strawberry import ModelInstance
from .strawberry.mutation.input import Input
from .strawberry.mutation.types import (
    ErrorType,
    RequiredToManyUpdateInput,
    RequiredToOneInput,
    ToManyCreateInput,
    ToManyUpdateInput,
    ToOneInput,
    ValidationErrorType,
)
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository
from .validation.base import InputValidationError

__all__ = (
    "ErrorType",
    "Input",
    "InputValidationError",
    "ModelInstance",
    "QueryHook",
    "RequiredToManyUpdateInput",
    "RequiredToOneInput",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemyConfig",
    "StrawchemySyncRepository",
    "ToManyCreateInput",
    "ToManyUpdateInput",
    "ToOneInput",
    "ValidationErrorType",
)
