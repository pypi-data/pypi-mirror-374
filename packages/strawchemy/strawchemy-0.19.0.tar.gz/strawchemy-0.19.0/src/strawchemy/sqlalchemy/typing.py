from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections import OrderedDict
    from collections.abc import Callable

    from sqlalchemy import Function
    from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session
    from sqlalchemy.orm import DeclarativeBase, Session, scoped_session
    from sqlalchemy.sql import SQLColumnExpression
    from strawchemy.strawberry.dto import OrderByEnum
    from strawchemy.strawberry.filters.base import GraphQLComparison

    from ._executor import QueryExecutor
    from .hook import QueryHook


__all__ = (
    "AnyAsyncSession",
    "AnySession",
    "AnySyncSession",
    "DeclarativeSubT",
    "DeclarativeT",
    "FunctionGenerator",
    "QueryExecutorT",
    "QueryHookCallable",
    "RelationshipSide",
    "SessionT",
    "StatementType",
)

DeclarativeT = TypeVar("DeclarativeT", bound="DeclarativeBase")
DeclarativeSubT = TypeVar("DeclarativeSubT", bound="DeclarativeBase")
QueryHookDeclarativeT = TypeVar("QueryHookDeclarativeT", bound="DeclarativeBase")
SessionT = TypeVar("SessionT", bound="AnySession")
QueryExecutorT = TypeVar("QueryExecutorT", bound="QueryExecutor[Any]")

RelationshipSide: TypeAlias = Literal["parent", "target"]
StatementType = Literal["lambda", "select"]
FunctionGenerator: TypeAlias = "Callable[..., Function[Any]]"
QueryHookCallable: TypeAlias = "QueryHook[QueryHookDeclarativeT]"
FilterMap: TypeAlias = "OrderedDict[tuple[type[Any], ...], type[GraphQLComparison]]"
AnySyncSession: TypeAlias = "Session | scoped_session[Session]"
AnyAsyncSession: TypeAlias = "AsyncSession | async_scoped_session[AsyncSession]"
AnySession: TypeAlias = "AnySyncSession | AnyAsyncSession"
OrderBySpec: TypeAlias = "tuple[SQLColumnExpression[Any], OrderByEnum]"
