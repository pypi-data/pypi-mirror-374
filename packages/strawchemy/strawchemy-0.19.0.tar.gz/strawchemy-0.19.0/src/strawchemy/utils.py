from __future__ import annotations

import inspect
import re
import sys
from typing import TYPE_CHECKING, Any, Optional, Union, get_args, get_origin

from strawchemy.typing import UNION_TYPES

if TYPE_CHECKING:
    from re import Pattern

__all__ = (
    "camel_to_snake",
    "is_type_hint_optional",
    "non_optional_type_hint",
    "snake_keys",
    "snake_to_camel",
    "snake_to_lower_camel_case",
)

_camel_to_snake_pattern: Pattern[str] = re.compile(r"((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))")


def camel_to_snake(string: str) -> str:
    """Convert a camelcased string to snake case.

    See: https://stackoverflow.com/a/12867228
    """
    return _camel_to_snake_pattern.sub(r"_\1", string).lower()


def snake_to_camel(string: str) -> str:
    """Convert string to camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    return "".join(x.capitalize() for x in string.lower().split("_"))


def snake_to_lower_camel_case(snake_str: str) -> Any:
    """Convert string to lower camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    camel_string: str = snake_to_camel(snake_str)
    return snake_str[0].lower() + camel_string[1:]


def snake_keys(value: dict[str, Any]) -> dict[str, Any]:
    """Recursively convert dict keys to from camel case to snake case."""
    res: dict[Any, Any] = {}
    for k, v in value.items():
        to_snake: str = camel_to_snake(k)
        if isinstance(v, (list, tuple)):
            res[to_snake] = [snake_keys(el) for el in v]
        elif isinstance(v, dict):
            res[to_snake] = snake_keys(v)
        else:
            res[to_snake] = v
    return res


def non_optional_type_hint(type_hint: Any) -> Any:
    origin, args = get_origin(type_hint), get_args(type_hint)
    if origin is Optional:
        return args
    if origin in UNION_TYPES:
        union_args = tuple([arg for arg in args if arg not in (None, type(None))])
        if len(union_args) == 1:
            return union_args[0]
        return Union[union_args]
    return type_hint


def is_type_hint_optional(type_hint: Any) -> bool:
    """Whether the given type hint is considered as optional or not.

    Returns:
        `True` if arguments of the given type hint are optional

    Three cases are considered:
    ```
        Optional[str]
        Union[str, None]
        str | None
    ```
    In any other form, the type hint will not be considered as optional
    """
    origin = get_origin(type_hint)
    if origin is None:
        return False
    if origin is Optional:
        return True
    if origin in UNION_TYPES:
        args = get_args(type_hint)
        return any(arg is type(None) for arg in args)
    return False


def get_annotations(obj: Any) -> dict[str, Any]:
    """Get the annotations of the given object."""
    if sys.version_info >= (3, 10):
        return inspect.get_annotations(obj)
    return obj.__dict__.get("__annotations__", {}) if isinstance(obj, type) else getattr(obj, "__annotations__", {})
