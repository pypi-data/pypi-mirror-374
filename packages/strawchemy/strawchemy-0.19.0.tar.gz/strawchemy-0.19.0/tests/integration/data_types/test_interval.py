from __future__ import annotations

from datetime import timedelta
from typing import TYPE_CHECKING, Any

import msgspec
import pytest

from sqlalchemy import Insert, MetaData, insert
from syrupy.assertion import SnapshotAssertion
from tests.integration.fixtures import QueryTracker
from tests.integration.models import IntervalModel, interval_metadata
from tests.integration.types import mysql as mysql_types
from tests.integration.types import postgres as postgres_types
from tests.integration.types import sqlite as sqlite_types
from tests.integration.typing import RawRecordData
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

if TYPE_CHECKING:
    from strawchemy.typing import SupportedDialect


seconds_in_year = 60 * 60 * 24 * 365.25
seconds_in_month = seconds_in_year / 12
seconds_in_day = 60 * 60 * 24


@pytest.fixture
def metadata() -> MetaData:
    return interval_metadata


@pytest.fixture
def seed_insert_statements(raw_intervals: RawRecordData) -> list[Insert]:
    return [insert(IntervalModel).values(raw_intervals)]


@pytest.fixture
def async_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.IntervalAsyncQuery
    if dialect == "mysql":
        return mysql_types.IntervalAsyncQuery
    if dialect == "sqlite":
        return sqlite_types.IntervalAsyncQuery
    pytest.skip(f"Interval tests can't be run on this dialect: {dialect}")


@pytest.fixture
def sync_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.IntervalSyncQuery
    if dialect == "mysql":
        return mysql_types.IntervalSyncQuery
    if dialect == "sqlite":
        return sqlite_types.IntervalSyncQuery
    pytest.skip(f"Interval tests can't be run on this dialect: {dialect}")


@pytest.mark.parametrize(
    ("component", "value", "expected_ids"),
    [
        pytest.param("days", timedelta(weeks=1, days=3, hours=12).total_seconds() / seconds_in_day, [1], id="days"),
        pytest.param("hours", timedelta(weeks=1, days=3, hours=12).total_seconds() / 3600, [1], id="hours"),
        pytest.param("minutes", timedelta(weeks=1, days=3, hours=12).total_seconds() / 60, [1], id="minutes"),
        pytest.param("seconds", timedelta(weeks=1, days=3, hours=12).total_seconds(), [1], id="totalSeconds"),
    ],
)
@pytest.mark.snapshot
async def test_timedelta_components(
    component: str,
    value: int,
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_intervals: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    query = f"""
        {{
            intervals(filter: {{ timeDeltaCol: {{ {component}: {{ eq: {value} }} }} }}) {{
                id
                timeDeltaCol
            }}
        }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["intervals"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["intervals"][i]["id"] == raw_intervals[expected_id]["id"]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_timedelta_output(
    any_query: AnyQueryExecutor,
    raw_intervals: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    query = """
        {
            intervals {
                id
                timeDeltaCol
            }
        }
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["intervals"]) == len(raw_intervals)

    for interval in result.data["intervals"]:
        expected_interval = next(f for f in raw_intervals if f["id"] == interval["id"])
        assert interval["timeDeltaCol"] == msgspec.json.encode(expected_interval["time_delta_col"]).decode()

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
