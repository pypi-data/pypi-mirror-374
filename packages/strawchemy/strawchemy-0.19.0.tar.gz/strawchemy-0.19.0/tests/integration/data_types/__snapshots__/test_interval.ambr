# serializer version: 1
# name: test_timedelta_components[session-tracked-async-days-aiosqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-days-asyncmy_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE unix_timestamp(interval_model.time_delta_col) / %s = %s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-days-asyncpg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST($1::INTEGER AS NUMERIC) = $2::NUMERIC
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-days-psycopg_async_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-hours-aiosqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-hours-asyncmy_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE unix_timestamp(interval_model.time_delta_col) / %s = %s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-hours-asyncpg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST($1::INTEGER AS NUMERIC) = $2::NUMERIC
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-hours-psycopg_async_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-minutes-aiosqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-minutes-asyncmy_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE unix_timestamp(interval_model.time_delta_col) / %s = %s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-minutes-asyncpg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST($1::INTEGER AS NUMERIC) = $2::NUMERIC
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-minutes-psycopg_async_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-totalSeconds-aiosqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-totalSeconds-asyncmy_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE unix_timestamp(interval_model.time_delta_col) = %s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-totalSeconds-asyncpg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) = $1::FLOAT
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-async-totalSeconds-psycopg_async_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) = %(param_1)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-days-psycopg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-days-sqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-hours-psycopg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-hours-sqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-minutes-psycopg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) / CAST(%(param_1)s::INTEGER AS NUMERIC) = %(param_2)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-minutes-sqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) / (? + 0.0) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-totalSeconds-psycopg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE EXTRACT(EPOCH FROM interval_model.time_delta_col) = %(param_1)s
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_components[session-tracked-sync-totalSeconds-sqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   WHERE CAST(strftime(?, interval_model.time_delta_col) AS INTEGER) = ?
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-async-aiosqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-async-asyncmy_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-async-asyncpg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-async-psycopg_async_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-sync-psycopg_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
# name: test_timedelta_output[session-tracked-sync-sqlite_engine]
  '''
  SELECT interval_model.time_delta_col,
         interval_model.id
    FROM interval_model AS interval_model
   ORDER BY interval_model.id ASC
  '''
# ---
